app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/WifiDirectBroadcastReceiver.java

```diff
@@ -0,0 +1,86 @@
+import android.content.Context;
+import android.content.Intent;
+import android.net.NetworkInfo;
+import android.net.wifi.p2p.WifiP2pDevice;
+import android.net.wifi.p2p.WifiP2pManager;
+import android.util.Log;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+/**
+ * Helper class for the local file sharing module.
+ *
+ * Handles the broadcasts pertaining to the wifi p2p group formed in WiFi Direct. Works along with
+ * the wifi p2p manager in {@link LocalFileTransferActivity}.
+ */
+public class WifiDirectBroadcastReceiver extends BroadcastReceiver {
+
+  private WifiP2pManager manager;
+  private WifiP2pManager.Channel channel;
+  private WifiDirectManager wifiDirectManager;
+
+  public WifiDirectBroadcastReceiver(@NonNull WifiP2pManager manager, @NonNull WifiP2pManager.Channel channel,
+      @NonNull WifiDirectManager wifiDirectManager) {
+    super();
+    this.manager = manager;
+    this.channel = channel;
+    this.wifiDirectManager = wifiDirectManager;
+  }
+
+  @Override
+  public void onReceive(@NonNull Context context, @NonNull Intent intent) {
+    String action = intent.getAction();
+
+    if (WifiP2pManager.WIFI_P2P_STATE_CHANGED_ACTION.equals(action)) {
+      // Update wifi p2p state
+      int wifiP2pState = intent.getIntExtra(WifiP2pManager.EXTRA_WIFI_STATE, -1);
+
+      if (wifiP2pState == WifiP2pManager.WIFI_P2P_STATE_ENABLED) {
+        ((BroadcastListener) wifiDirectManager).setWifiP2pEnabled(true);
+      } else {
+        ((BroadcastListener) wifiDirectManager).setWifiP2pEnabled(false);
+      }
+      Log.d(LocalFileTransferActivity.TAG, "WiFi P2P state changed - " + wifiP2pState);
+
+
+    } else if (WifiP2pManager.WIFI_P2P_PEERS_CHANGED_ACTION.equals(action)) {
+      if (manager != null) {
+        /* List of available peers has changed, so request & use the new list through
+         * PeerListListener.requestPeers() callback */
+        manager.requestPeers(channel, wifiDirectManager);
+      }
+      Log.d(LocalFileTransferActivity.TAG, "P2P peers changed");
+
+
+    } else if (WifiP2pManager.WIFI_P2P_CONNECTION_CHANGED_ACTION.equals(action)) {
+      if (manager == null) {
+        return;
+      }
+      NetworkInfo networkInfo = intent.getParcelableExtra(WifiP2pManager.EXTRA_NETWORK_INFO);
+
+      if (networkInfo.isConnected()) {
+        // Request connection info about the wifi p2p group formed upon connection
+        manager.requestConnectionInfo(channel, wifiDirectManager);
+      } else {
+        // Not connected after connection change -> Disconnected
+        ((BroadcastListener) wifiDirectManager).onDisconnected();
+      }
+
+
+    } else if (WifiP2pManager.WIFI_P2P_THIS_DEVICE_CHANGED_ACTION.equals(action)) {
+      ((BroadcastListener) wifiDirectManager).onDeviceChanged(intent.getParcelableExtra(WifiP2pManager.EXTRA_WIFI_P2P_DEVICE));
+    }
+  }
+
+  public interface BroadcastListener {
+    void setWifiP2pEnabled(boolean wifiP2pEnabled);
```

Comment:
A listener should emit Events, in one of my comments on this file I gave an example for this specific function
```
listener.onWifiP2PStateChanged(wifiP2pState == WifiP2pManager.WIFI_P2P_STATE_ENABLED)
```
and the definition would be
```
void onWifiP2PStateChanged(boolean isEnabled)
```
by saying `set` you are more commanding an action but this class should know absolutely nothing about how classes will use the information it reports, it should just report information via a listener.

`BroadcastListener` is also a bit of a weak name, generally I am for more descriptiveness, what this is a listener for is peer to peer change events so full descriptiveness would demand "P2PStateChangeListener" or the other route is to just call it listener and be fine with the qualified access describing it eg "WifiDirectBroadcastReceiver.Listener".
On the topic of naming `WifiDirectBroadcastReceiver` could use an improvement too because it is definitely a WifiP2PBroadcastReceiver

Reply:
> `void onWifiP2PStateChanged(boolean isEnabled)`

I'll update this

> BroadcastListener is also a bit of a weak name, ....  definitely a WifiP2PBroadcastReceiver

The reason why I stayed clear of putting P2p in the name is because most of the P2P API methods and classes are of the form `WifiP2p...`, so I didn't want to cause confusion about the code

Since our code is using `WifiDirect...`, how about  leaving WifiDirectBroadcastReceiver as is and changing BroadcastListener to `WifiDirectBroadcastReceiver.EventListener` as suggested?

Reply:
When I stumble on a name and can't think of something better I prefix it with the name of the app I am working on, it is not the best but avoids the confusion eg `KiwixWifiP2PBroadcastListener`.
I definitely like to keep `Wifip2p` in there somewhere because it better describes what this class is actually doing, WifiDirectManager should stay the same it is about where the class is used that is important.
From the Activity I know it is a file transfer activity and by looking at wifidirectmanager I see that it uses wifidirect and that is all I care about, if I work on WifiDirectManager it is now more important that I know how it works and a name with P2P is a good clue with the current level of abstraction

Reply:
So `KiwixWifiP2pBroadcastListener` and `KiwixWifiP2pBroadcastListener.P2pEventListener` work?

Reply:
Yes that seems good to me

Reply:
[Done](https://github.com/kiwix/kiwix-android/pull/1276/commits/a00481bb5b88ef594c6fa83ed95660129be56a84)
Will fix the CodeFactor check issues in the next commit

Reply:
it has not been renamed onWifiP2PStateChanged and you are using the values `true` & `false` when the boolean condition in the if statement is the argument you want to supply to the callback.
Manager and channel should not be provided in the constructor, if you need to perform an action with the fields of WDM then use a callback

Reply:
[Done](https://github.com/kiwix/kiwix-android/pull/1276/commits/ba00ece2ca1b9d3cba9ee40eba1c0a2b2022c8fd#diff-e84e03042026401a0660fa610753ae45R59) now
I'm passing the state value (an int) itself in the callback, using the if statement at implementation side
Also removed manager and channel